### PR TITLE
Text-to-Image Generation Demo with Dall-E Mini

### PR DESCRIPTION
**Pull request by aryan1107**
For task #767 
A demo for text to image model:

Screen Shot 2022-08-26 at 2 25 44 AM

Screen Shot 2022-08-19 at 12 36 50 AM

### CODE REVIEW

File: lit_nlp/examples/models/dalle.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    def p_decode(indices, params):
+        return vqgan.decode_code(indices, params=params)
+    
+    # create a random key
+    seed = random.randint(0, 2**32 - 1)
+    key = jax.random.PRNGKey(seed)
+    
+    prompts = [ex["prompt"] for ex in inputs]
+
+    processor = DalleBartProcessor.from_pretrained(DALLE_MODEL, revision=DALLE_COMMIT_ID)
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:103]: This seems like it should be initialized in `__init__()` as a property of the `DalleModel` class.

File: lit_nlp/examples/models/dalle.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    tokenized_prompt = replicate(tokenized_prompts)
+
+    # generate Images
+    # number of predictions per prompt
+    n_predictions = self.predictions
+    # We can customize generation parameters ( https://huggingface.co/blog/how-to-generate)
+    gen_top_k = None
+    gen_top_p = None
+    temperature = None
+    cond_scale = 10.0
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:114]: These could use better type annotations (especially the `gen_*` properties, can't tell if they're `bool`, `int`, or something else), but also seem like great candidates for `Optional` fields in the `input_spec()`. Will be important to get the validation right for these though (e.g., can you use temp with top_k, etc.).

File: lit_nlp/examples/models/dalle.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    
+    prompts = [ex["prompt"] for ex in inputs]
+
+    processor = DalleBartProcessor.from_pretrained(DALLE_MODEL, revision=DALLE_COMMIT_ID)
+    tokenized_prompts = processor(prompts)
+    tokenized_prompt = replicate(tokenized_prompts)
+
+    # generate Images
+    # number of predictions per prompt
+    n_predictions = self.predictions
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:109]: Rename `self.predictions` as `self.n_predictions` or similar as the former has a certain connotation in LIT (i.e., the literal predictions output from the model) and we don't want to overload that namespace.

File: lit_nlp/examples/models/dalle.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            cond_scale,
+        )
+        # remove BOS
+        encoded_images = encoded_images.sequences[..., 1:]
+        # decode images
+        decoded_images = p_decode(encoded_images, vqgan_params)
+        decoded_images = decoded_images.clip(0.0, 1.0).reshape((-1, 256, 256, 3))
+        for decoded_img in decoded_images:
+            img = Image.fromarray(np.asarray(decoded_img * 255, dtype=np.uint8))
+            images.append(img)
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:139]: What's happening with this list?

File: lit_nlp/examples/models/dalle.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    gen_top_k = None
+    gen_top_p = None
+    temperature = None
+    cond_scale = 10.0
+
+    # generate images
+    images = []
+    final_Output = []
+    for i in trange(max(n_predictions // jax.device_count(), 1)):
+        # get a new key
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:120]: Enrich this comment or add a new one with more details about why we need this key to do a random split, and how this impacts generation.

File: lit_nlp/examples/datasets/dalle_prompt.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  def __init__(self):
+    """Dataset constructor, loads the data into memory."""
+    data = {'prompt': ["Batman feeding his pet Goldfish"," You're at a museum and there's a painting of a woman wearing a white dress. She has a blue scarf around her neck and she's holding a red rose. You take a picture of the painting, but when you look at the picture, the woman is gone and there's a blue bird in her place.","a kiwi that is both a bird and a fruit"]}
+    # into datafrane to decode string
+    df = pd.DataFrame(data=data)
+
+    # populate this with data records
+    self._examples = [{
+      'prompt': row['prompt'],
+    } for _, row in df.iterrows()]
```
Comment (RyanMullins) [inline lit_nlp/examples/datasets/dalle_prompt.py:19]: You could just do a comprehension over the list of input strings instead of using a DataFrame here.

File: lit_nlp/examples/datasets/dalle_prompt.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+from lit_nlp.api import dataset as lit_dataset
+from lit_nlp.api import types as lit_types
+import pandas as pd
+
+class Dalle(lit_dataset.Dataset):
+  """TyDiQA dataset."""
+
+  def __init__(self):
+    """Dataset constructor, loads the data into memory."""
+    data = {'prompt': ["Batman feeding his pet Goldfish"," You're at a museum and there's a painting of a woman wearing a white dress. She has a blue scarf around her neck and she's holding a red rose. You take a picture of the painting, but when you look at the picture, the woman is gone and there's a blue bird in her place.","a kiwi that is both a bird and a fruit"]}
```
Comment (RyanMullins) [inline lit_nlp/examples/datasets/dalle_prompt.py:12]: We enforce an 80 character [line length](https://google.github.io/styleguide/pyguide.html#32-line-length) when possible. For this, any individual input string can exceed the 80 character limit but each input string should be on its own line.

File: environment.yml [hunk 8]
```diff
@@ -44,3 +44,10 @@ dependencies:
         - Jinja2
         - stanza
         - shap==0.37.0
+        - jax
+        - jaxlib
+        - flax
+        - dalle-mini
+        - git+https://github.com/patil-suraj/vqgan-jax.git
```
Comment (RyanMullins) [inline environment.yml:51]: @jameswex do we have any bets practices about importing repos from GitHub like this?

File: lit_nlp/examples/dalle_demo.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+# Custom frontend layout; see api/layout.py
+modules = layout.LitModuleName
+Dalle_LAYOUT = layout.LitCanonicalLayout(
+    upper={
+        "Main": [
+            modules.EmbeddingsModule,
+            modules.DataTableModule,
+            modules.DatapointEditorModule,
+            modules.SliceModule,
+            modules.ColorModule,
```
Comment (RyanMullins) [inline lit_nlp/examples/dalle_demo.py:47]: Could probably get away with just the `DataTableModule` and `DatapointEditorModule` modules, but if stuff shows up in the others we can keep them.

File: lit_nlp/examples/models/dalle.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
+"""Dalle model based on https://github.com/borisdayma/dalle-mini."""
+from ast import Str
+import re
+from typing import List
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:4]: Unused import?

File: lit_nlp/examples/models/dalle.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
+"""Dalle model based on https://github.com/borisdayma/dalle-mini."""
+from ast import Str
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:2]: Prefer the built-in `str` primitive.

File: lit_nlp/examples/models/dalle.py (hunk 3 above)
Comment (RyanMullins) [reply lit_nlp/examples/models/dalle.py:109]: Also, if we want to suppor omre than 1 predicted image, we'll need to update a few other things:

- [ ] Add a `ImageBytesList` type to types.py (but need to wait for Crystal's type refactor PRs before adding this in this PR)
- [ ] Update [`GeneratedImageModule.supportedTypes`](https://github.com/PAIR-code/lit/blob/dev/lit_nlp/client/modules/generated_image_module.ts#L44) to include the `ImageBytesList`
- [ ] Update [`GeneratedImageModule.renderImage(...)`](https://github.com/PAIR-code/lit/blob/dev/lit_nlp/client/modules/generated_image_module.ts#L88) to accept `src: string | string[]` and use `Array.isArray(...)` to determine if we should use `src.map(...)` to add the `` elements or if we can just use the value of `src` directly.
- [ ] Probably need to update some CSS for those `` elements, too.

File: lit_nlp/examples/models/dalle.py (hunk 4 above)
Comment (aryan1107) [reply lit_nlp/examples/models/dalle.py:139]: Honestly, i don't understand but I followed documentation they did it this way and it said: 

"Model functions are compiled and parallelized to take advantage of multiple devices." 

I just used same thing.

File: lit_nlp/examples/datasets/dalle_prompt.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+"""Data loaders for text-to-image model."""
+
+from lit_nlp.api import dataset as lit_dataset
+from lit_nlp.api import types as lit_types
+import pandas as pd
+
+class Dalle(lit_dataset.Dataset):
+  """TyDiQA dataset."""
```
Comment (RyanMullins) [inline lit_nlp/examples/datasets/dalle_prompt.py:8]:
```suggestion
  """Example prompts dataset for use with Dall-E and other text-to-image generative models."""
```

File: lit_nlp/examples/dalle_demo.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    "Models to load")
+
+_MAX_EXAMPLES = flags.DEFINE_integer(
+    "max_examples", 10,
+    "Maximum number of examples to load from each evaluation set. Set to None to load the full set."
+)
+
+# Custom frontend layout; see api/layout.py
+modules = layout.LitModuleName
+Dalle_LAYOUT = layout.LitCanonicalLayout(
```
Comment (RyanMullins) [inline lit_nlp/examples/dalle_demo.py:40]: Use `DALLE_LAYOUT` instead, to keep with the style guide for module constant naming.

File: lit_nlp/examples/dalle_demo.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    upper={
+        "Main": [
+            modules.EmbeddingsModule,
+            modules.SliceModule,
+            modules.ColorModule,
```
Comment (RyanMullins) [inline lit_nlp/examples/dalle_demo.py:45]:
```suggestion
            modules.DatapointEditor
```

We always want the DatapointEditor in layouts so that users can add new data, and the Color Module is not long for this earth.

File: lit_nlp/examples/dalle_demo.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    "max_examples", 10,
+    "Maximum number of examples to load from each evaluation set. Set to None to load the full set."
+)
+
+# Custom frontend layout; see api/layout.py
+modules = layout.LitModuleName
+Dalle_LAYOUT = layout.LitCanonicalLayout(
+    upper={
+        "Main": [
+            modules.EmbeddingsModule,
```
Comment (RyanMullins) [inline lit_nlp/examples/dalle_demo.py:43]: I like the idea of supporting embeddings in this demo. Would be really cool if we had embeddings for the prompts and the images so that we could toggle between these fields, but that sounds somewhat complicated since the output will be multiple images and each will require their own embedding.

File: lit_nlp/examples/models/dalle.py (hunk 2 above)
Comment (RyanMullins) [reply lit_nlp/examples/models/dalle.py:114]: Can use the types from https://github.com/borisdayma/dalle-mini/blob/main/src/dalle_mini/model/modeling.py#L1589

```suggestion
    gen_top_k: Optional[int] = None
    gen_top_p: Optional[float] = None
    temperature: Optional[float] = None
    cond_scale: Optional[float] = 10.0
```

File: lit_nlp/examples/models/dalle.py (hunk 3 above)
Comment (RyanMullins) [reply lit_nlp/examples/models/dalle.py:109]: Cleared the implement the above in this PR. 

Make sure to pull the latest from https://github.com/PAIR-code/lit/commits/dev into your branch before starting on this. 

Need to add a `ImageBytesList` class in two places: first in [types.py](https://github.com/PAIR-code/lit/blob/dev/lit_nlp/api/types.py) then in [lit_types.ts](https://github.com/PAIR-code/lit/blob/dev/lit_nlp/client/lib/lit_types.ts). Both should inherit from `_List`, but can otherwise be empty. Be sure the use the `@registered` decorator with the TypeScript class, this is required so that we can process model and dataset `Spec` information. 

Hopefully the rest goes down smoothly. Let me know if you have any questions.

File: lit_nlp/examples/models/dalle.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+            # Append CLIP score depending on prompt size hence the condition
+            if len(np.shape(logits)) == 1:
+                clip_score.append((str(logits[idx]), None))
+            else:
+                clip_score.append((str(logits[i][idx]), None))
+        # Append to final List[Dict]
+        final_images.append({
+            "image": images_per_prompt,
+            "clip_score": clip_score
+            })
```
Comment (RyanMullins) [inline lit_nlp/examples/models/dalle.py:213]:
```suggestion
        })
```

File: lit_nlp/examples/models/dalle.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+    clip_repo = "openai/clip-vit-base-patch32"
+    clip_commit_id = None
+
+    # Load CLIP
+    self.clip, clip_params = FlaxCLIPModel.from_pretrained(
+        clip_repo, revision=clip_commit_id, dtype=jnp.float16, _do_init=False
+    )
+
+    self.clip_processor = CLIPProcessor.from_pretrained(clip_repo, revision=clip_commit_id)
+    self.clip_params = replicate(clip_params)
```
[the remark this answers is not in the document]
Comment (aryan1107) [reply lit_nlp/examples/models/dalle.py]: Wow this is neat! Don't know why couldn't I come up with this at first.

File: lit_nlp/examples/models/dalle_test.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    ospec = model.output_spec()
+    assert "image" in ospec
+    assert isinstance(
+        ospec["image"],
+        (lit_types.ImageBytesList))
+    assert "clip_score" in ospec
+    assert isinstance(
+        ospec["clip_score"],
+        (lit_types.GeneratedTextCandidates))
+    assert ospec["clip_score"].parent == "prompt"
```
[the remark this answers is not in the document]
Comment (aryan1107) [reply lit_nlp/examples/models/dalle_test.py]: Thanks, Will also update TyDi test with [absltest library](https://abseil.io/docs/python/guides/testing)

File: lit_nlp/api/types.py [hunk 19]
```diff
@@ -141,6 +141,11 @@ class ListLitType(LitType):
   default: Sequence[Any] = None
 
 
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class ImageBytesList(ListLitType):
+  """A list of strings."""
```
Comment (RyanMullins) [inline lit_nlp/api/types.py:146]: Fix comment.

**Review by RyanMullins — APPROVED** [review]

File: lit_nlp/api/types.py (hunk 19 above)
Comment (aryan1107) [reply lit_nlp/api/types.py:146]: How about `A list of ImageBytes` ?

File: environment.yml (hunk 8 above)
Comment (jameswex) [reply environment.yml:52]: haven't had to do it before, but would be best to tag a specific commit/version to avoid breakage.

**Comment by RyanMullins** [comment]
After further inspection, I found that more recent versions of `dalle-mini` pin to a specific version of `transformers` (v4.25.1), which is behind the rest of LIT's examples but compatible with the core library. Rather than deal with upstream dependency issues, I've isolated the DALL-E Mini demo to its own directory, removed tests that would otherwise conflict with our test suite, and written a stand-along requirements file and README with installation instructions. 

If the CI passes, I'll approve this PR and we can merge it into `dev`.

**Review by RyanMullins — APPROVED** [review]

**Review by RyanMullins — APPROVED** [review]

**Review by RyanMullins — APPROVED** [review]
